### PR TITLE
Add spec for governance token staking from general account and locked rewards counting towards staking balance

### LINKED ISSUE #2347: Governance token staking from general account and locked rewards counting towards staking balance
# Governance token staking from general account

A new account to be created `locked_for_staking`. It should be possible to transfer the governance token in and out of this account. Balance in this account should count towards tokens "associated" to the vega key for staking purposes and can be staked to a specific validators, thus earning staking rewards for the key. 

#  Locked rewards counting towards staking balance

Any rewards paid in the governance token that are "locked" should count towards the staking balance as "associated" to the vega key for staking purposes and can be staked to a specific validators, thus earning staking rewards for the key.

### PR DESCRIPTION
**Pull request by Jiajia-Cui**
closes #2347

### CODE REVIEW

File: protocol/0013-ACCT-accounts.md [hunk 1]
```diff
@@ -2,8 +2,11 @@
 
 ## Accounts controlled by parties
 
-A party only has control over balances in the "general" account for each asset.
-[Parties](./0017-PART-party.md) are identified by Vega public keys. Each party that makes a deposit on one of the asset bridges, currently only [Ethereum ERC20 bridge](./0031-ETHB-ethereum_bridge_spec.md) will have a general account for the relevant [asset](./0040-ASSF-asset_framework.md) created.
+A party only has control over balances in the "general" and "LOCKED_FOR_STAKING" account for each asset.
+[Parties](./0017-PART-party.md) are identified by Vega public keys. Each party that makes a deposit on one of the asset bridges, currently only [Ethereum ERC20 bridge](./0031-ETHB-ethereum_bridge_spec.md) will have a general account and "LOCKED_FOR_STAKING" for the relevant [asset](./0040-ASSF-asset_framework.md) created.
+
+A party can transfer governance tokens in and out of the "LOCKED_FOR_STAKING" account. The balance in this account should count towards the tokens "associated" with the Vega key for staking purposes and can be staked to specific validators, earning staking rewards for that key.
```
Comment (davidsiska-vega) [inline protocol/0013-ACCT-accounts.md]:
```suggestion
A party can transfer governance tokens in and out of the "LOCKED_FOR_STAKING" account. The balance in this account should count towards the tokens [associated](./0059-STKG-simple_staking_and_delegating.md) with the Vega key for staking purposes and can be staked to specific validators, earning staking rewards for that key.
```

File: protocol/0057-TRAN-transfers.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ These transfers are not to be confused with the internal concept of transfers wh
 Allowing users to initiate transfers allows for the following capabilities:
 
 - A user can transfer funds from a public key A to a public key B.
-- A user can transfer funds from and to a locked account used for staking (yet to be specified) [LOCKED_FOR_STAKING](./0059-STKG-simple_staking_and_delegating.md).
+- A user can transfer funds from and to a locked account used for staking [LOCKED_FOR_STAKING](./0059-STKG-simple_staking_and_delegating.md).
```
Comment (davidsiska-vega) [inline protocol/0057-TRAN-transfers.md]:
```suggestion
- A user can transfer the governance token only from and to a locked account used for staking [LOCKED_FOR_STAKING](./0059-STKG-simple_staking_and_delegating.md).
```

File: protocol/0057-TRAN-transfers.md [hunk 3]
```diff
@@ -259,8 +259,8 @@ message CancelTransfer {
 
 - As a user I can transfer funds from a general account I control to an other party's general account. Such transfer can be immediate or delayed. (<a name="0057-TRAN-001" href="#0057-TRAN-001">0057-TRAN-001</a>)
 - As a user I **cannot** transfer funds from a general account I control to reward account with a one-off transfer. (<a name="0057-TRAN-002" href="#0057-TRAN-002">0057-TRAN-002</a>)
-- As a user I can transfer funds from a general account I control to a locked_for_staking. Such transfer can be immediate or delayed. This functionality is currently not implemented (so don't try to test) (<a name="0057-PALAZZO-003" href="#0057-PALAZZO-003">0057-PALAZZO-003</a>).
-- As a user I can transfer funds from a locked_from_staking account under my control to any party's general_account. Such transfer can be immediate or delayed. This functionality is currently not implemented (so don't try to test) (<a name="0057-PALAZZO-004" href="#0057-PALAZZO-004">0057-PALAZZO-004</a>)
+- As a user I can transfer funds from a general account I control to a LOCKED_FOR_STAKING account. Such transfer can be immediate or delayed. (<a name="0057-TRAN-003" href="#0057-TRAN-003">0057-TRAN-003</a>).
```
Comment (davidsiska-vega) [inline protocol/0057-TRAN-transfers.md]:
```suggestion
- As a user I can transfer the governance token from a general account I control to a LOCKED_FOR_STAKING account. Such transfer must be immediate. (<a name="0057-TRAN-003" href="#0057-TRAN-003">0057-TRAN-003</a>).
```

File: protocol/0057-TRAN-transfers.md [hunk 4]
```diff
@@ -259,8 +259,8 @@ message CancelTransfer {
...
+- As a user I can transfer funds from a LOCKED_FOR_STAKING account under my control to any party's general account. Such transfer can be immediate or delayed.  (<a name="0057-TRAN-004" href="#0057-TRAN-004">0057-TRAN-004</a>)
```
Comment (davidsiska-vega) [inline protocol/0057-TRAN-transfers.md]:
```suggestion
- As a user I can transfer the governance token from a LOCKED_FOR_STAKING account under my control to any party's general account. Such transfer must be immediate.  (<a name="0057-TRAN-004" href="#0057-TRAN-004">0057-TRAN-004</a>)
```

File: protocol/0059-STKG-simple_staking_and_delegating.md [hunk 5]
```diff
@@ -63,6 +63,8 @@ towards the next epoch, though it can be undone through undelegate.
 
 Once Vega is aware of locked tokens, the users will have an [account](./0013-ACCT-accounts.md#staking-accounts) with the balance reflecting how many tokens were locked. At this point, the user can submit a transaction to stake (delegate) their tokens. The amount they stake must be `<= balance`, naturally.
 
+Any rewards paid in governance tokens that are locked should be included in the same [account].
```
Comment (davidsiska-vega) [inline protocol/0059-STKG-simple_staking_and_delegating.md]: What does this mean?

Comment (davidsiska-vega) [reply protocol/0059-STKG-simple_staking_and_delegating.md]: Also we'd want ACs for locked rewards.

**Review by davidsiska-vega — CHANGES_REQUESTED** [review]
We need more detail around staking locked rewards.

File: protocol/0013-ACCT-accounts.md [hunk 6]
```diff
@@ -2,8 +2,13 @@
 
 ## Accounts controlled by parties
 
-A party only has control over balances in the "general" account for each asset.
-[Parties](./0017-PART-party.md) are identified by Vega public keys. Each party that makes a deposit on one of the asset bridges, currently only [Ethereum ERC20 bridge](./0031-ETHB-ethereum_bridge_spec.md) will have a general account for the relevant [asset](./0040-ASSF-asset_framework.md) created.
+A party only has control over balances in the "general" and "LOCKED_FOR_STAKING" account for each asset.
+[Parties](./0017-PART-party.md) are identified by Vega public keys. Each party that makes a deposit on one of the asset bridges, currently only [Ethereum ERC20 bridge](./0031-ETHB-ethereum_bridge_spec.md) will have a general account and "LOCKED_FOR_STAKING" for the relevant [asset](./0040-ASSF-asset_framework.md) created.
```
Comment (davidsiska-vega) [inline protocol/0013-ACCT-accounts.md]:
```suggestion
[Parties](./0017-PART-party.md) are identified by Vega public keys. Each party that makes a deposit on one of the asset bridges, currently only [Ethereum ERC20 bridge](./0031-ETHB-ethereum_bridge_spec.md) will have a general account for the relevant [asset](./0040-ASSF-asset_framework.md) created.
A party that transfers any balance of the governance token into  its "LOCKED_FOR_STAKING" account will have that account created.
```

File: protocol/0057-TRAN-transfers.md [hunk 7]
```diff
@@ -6,7 +6,9 @@ These transfers are not to be confused with the internal concept of transfers wh
 Allowing users to initiate transfers allows for the following capabilities:
 
 - A user can transfer funds from a public key A to a public key B.
-- A user can transfer funds from and to a locked account used for staking (yet to be specified) [LOCKED_FOR_STAKING](./0059-STKG-simple_staking_and_delegating.md).
+- A user can transfer the governance token only from and to a locked account used for staking [LOCKED_FOR_STAKING](./0059-STKG-simple_staking_and_delegating.md).
```
Comment (davidsiska-vega) [inline protocol/0057-TRAN-transfers.md]: This isn't true. You can also transfer vega tokens from general account to another key's general account.

File: protocol/0059-STKG-simple_staking_and_delegating.md [hunk 8]
```diff
@@ -18,6 +18,8 @@ nomination. To this end, a Vega token (or a fraction thereof) can be:
 - Unassociated: The tokenholder is free to do with the token as they want, but cannot nominate it
 - Associated: The token is locked in the staking and delegation smart contract and associated to a Vega key. It can be used on the Vega chain for governance and it can be nominated to a validator.
 
+Governance tokens locked in the "LOCKED_FOR_STAKING" account are associated with a VEGA key. To stake these locked governance tokens, the party must delegate them to one or more validators. When tokens are transferred in or out of the "LOCKED_FOR_STAKING" account, the keys associated with the tokens will be unassociated and associated to the new keys accordingly. However, the validator to whom the tokens were delegated will remain unchanged unless the new token owner decides to change the validator.
```
Comment (davidsiska-vega) [inline protocol/0059-STKG-simple_staking_and_delegating.md]:
```suggestion
Governance tokens locked in the "LOCKED_FOR_STAKING" account count as associated with the VEGA key of the party controlling the account.
To stake these locked governance tokens, the party must delegate them to one or more validators.

When governance token balance is transferred out of the "LOCKED_FOR_STAKING" account, the key's balance of associated token will be reduced accordingly and the staking balances removed from validators.
```

File: protocol/0013-ACCT-accounts.md [hunk 9]
```diff
@@ -2,8 +2,13 @@
...
+
+A party can transfer governance tokens in and out of the "LOCKED_FOR_STAKING" account. The balance in this account should count towards the tokens [associated](./0059-STKG-simple_staking_and_delegating.md) with the Vega key for staking purposes and can be staked to specific validators, earning staking rewards for that key.
+
+Vega core listens to events on other chains (currently Ethereum) for events on the bridge contract. If governance token reward is locked on the bridge contract with a Vega party key associated then the [Locked_FOR_REWARD](./0013-ACCT-accounts.md) balance for the party is incremented by the appropriate amount.
```
Comment (davidsiska-vega) [inline protocol/0013-ACCT-accounts.md]:
```suggestion
```

File: protocol/0013-ACCT-accounts.md [hunk 10]
```diff
@@ -182,3 +187,8 @@ One key difference with staking accounts is that the collateral is not held in a
 
 - It is possible to deposit funds from Ethereum directly into the global rewards account by specifying the `0` Vega address. (<a name="0013-ACCT-027" href="#0013-ACCT-027">0013-ACCT-027</a>)
 - It is possible to transfer funds from a Vega general account to the global rewards account by specifying the `0` address and appropriate account type. (<a name="0013-ACCT-028" href="#0013-ACCT-028">0013-ACCT-028</a>)
+
```
Comment (davidsiska-vega) [inline protocol/0013-ACCT-accounts.md:188]: AC for rewards are in a different place, I wouldn't randomly add them here.

**Review by davidsiska-vega — DISMISSED** [review]
more changes discussed

File: protocol/0056-REWA-rewards_overview.md [hunk 11]
```diff
@@ -403,6 +403,14 @@ Create 3 markets settling in USDT. Wait for a new epoch to begin, in the next ep
 
 Run for another epoch with no fee generated. Expect no transfer to be made to the reward pools of the accounts.
 
+### Locked Reward
+
+When a party earns rewards (in governance token), they are transferred into the party's locked account and locked for the appropriate period before vesting. Sum of rewards in vesting account and locked account will be reflected in the party's "LOCKED_FOR_STAKING" account and they can be staked.
```
Comment (davidsiska-vega) [inline protocol/0056-REWA-rewards_overview.md]:
```suggestion
When a party earns locked rewards (in governance token), they are transferred into the party's locked rewards account and locked for the appropriate period before vesting.
Sum of rewards in vesting account and locked account should count toward the party's staking balance and they can be [staked](0059-STKG-simple_staking_and_delegating.md).
```

File: protocol/0056-REWA-rewards_overview.md [hunk 12]
```diff
@@ -403,6 +403,14 @@ Create 3 markets settling in USDT. Wait for a new epoch to begin, in the next ep
...
+
+- When a party receives a governance token reward that is locked, the [Locked_FOR_REWARD] balance for the party is incremented by the corresponding amount. (<a name="0056-REWA-200" href="#0056-REWA-200">0056-REWA-200</a>).
+- When a party transfers locked tokens out of the [Locked_FOR_REWARD] account, the balance for the party is decreased by the corresponding amount. (<a name="0056-REWA-201" href="#0056-REWA-201">0056-REWA-201</a>).
```
Comment (davidsiska-vega) [inline protocol/0056-REWA-rewards_overview.md]:
```suggestion
```

File: protocol/0056-REWA-rewards_overview.md [hunk 13]
```diff
@@ -403,6 +403,14 @@ Create 3 markets settling in USDT. Wait for a new epoch to begin, in the next ep
...
+- A party can stake the assets in the [Locked_FOR_REWARD] account. (<a name="0056-REWA-202" href="#0056-REWA-202">0056-REWA-202</a>).
```
Comment (davidsiska-vega) [inline protocol/0056-REWA-rewards_overview.md]:
```suggestion
- Balance of the governance asset in the [Locked_FOR_REWARD] account counts towards the party's staking balance and can be nominated to a validator and earn staking rewards. (<a name="0056-REWA-202" href="#0056-REWA-202">0056-REWA-202</a>).
```

File: protocol/0059-STKG-simple_staking_and_delegating.md [hunk 14]
```diff
@@ -18,6 +18,10 @@ nomination. To this end, a Vega token (or a fraction thereof) can be:
 - Unassociated: The tokenholder is free to do with the token as they want, but cannot nominate it
 - Associated: The token is locked in the staking and delegation smart contract and associated to a Vega key. It can be used on the Vega chain for governance and it can be nominated to a validator.
 
+Governance tokens locked in the "LOCKED_FOR_STAKING" account count as associated with the VEGA key of the party controlling the account.
```
Comment (davidsiska-vega) [inline protocol/0059-STKG-simple_staking_and_delegating.md]:
```suggestion
Governance tokens that have been transferred into the "LOCKED_FOR_STAKING" account count as associated with the VEGA key of the party controlling the account.
```

File: protocol/0056-REWA-rewards_overview.md (hunk 12 above)
Comment (Jiajia-Cui) [reply protocol/0056-REWA-rewards_overview.md]: why delete this AC?

File: protocol/0013-ACCT-accounts.md [hunk 15]
```diff
@@ -2,8 +2,11 @@
 
 ## Accounts controlled by parties
 
-A party only has control over balances in the "general" account for each asset.
+A party only has control over balances in the "general" and "LOCKED_FOR_STAKING" account for each asset.
 [Parties](./0017-PART-party.md) are identified by Vega public keys. Each party that makes a deposit on one of the asset bridges, currently only [Ethereum ERC20 bridge](./0031-ETHB-ethereum_bridge_spec.md) will have a general account for the relevant [asset](./0040-ASSF-asset_framework.md) created.
+
+Each party that has locked and vesting rewards will have a "LOCKED_FOR_STAKING" account for the relevant [asset] created, party can transfer the tokens in and out of this LOCKED_FOR_STAKING account. The balance in this account should count towards the tokens [associated](./0059-STKG-simple_staking_and_delegating.md) with the Vega key for staking purposes and can be staked to specific validators, earning staking rewards for that key.
```
Comment (jeremyletang) [inline protocol/0013-ACCT-accounts.md]: I believe that the funds will be staying in the vesting / vested accounts, LOCKED_FOR_STAKING should be used only from a general account. 

The account will be created on demand if needed.

File: protocol/0056-REWA-rewards_overview.md [hunk 16]
```diff
@@ -403,6 +403,14 @@ Create 3 markets settling in USDT. Wait for a new epoch to begin, in the next ep
 
 Run for another epoch with no fee generated. Expect no transfer to be made to the reward pools of the accounts.
 
+### Locked Reward
+
+When a party earns locked rewards (in governance token), they are transferred into the party's locked rewards account and locked for the appropriate period before vesting.
+Sum of rewards in vesting account and locked account should count toward the party's staking balance and they can be [staked](0059-STKG-simple_staking_and_delegating.md).
```
Comment (cdummett) [inline protocol/0056-REWA-rewards_overview.md]: Small clarification, in cores implementation there are no separate locked rewards and vesting accounts, they are one single account.
```suggestion
When a party earns locked rewards (in governance token), they are transferred into the party's vesting account account but locked for the appropriate period before vesting.
The rewards in the vesting account (both locked and unlocked) should count toward the party's staking balance and they can be [staked](0059-STKG-simple_staking_and_delegating.md).
```

File: protocol/0056-REWA-rewards_overview.md [hunk 17]
```diff
@@ -403,6 +403,14 @@ Create 3 markets settling in USDT. Wait for a new epoch to begin, in the next ep
...
+
+- When a party receives a governance token reward that is locked, the [Locked_FOR_REWARD] balance for the party is incremented by the corresponding amount. (<a name="0056-REWA-200" href="#0056-REWA-200">0056-REWA-200</a>).
```
Comment (cdummett) [inline protocol/0056-REWA-rewards_overview.md]:
```suggestion
- When a party receives a governance token reward that is locked, the [LOCKED_FOR_STAKING] balance for the party is incremented by the corresponding amount. (<a name="0056-REWA-200" href="#0056-REWA-200">0056-REWA-200</a>).
```

File: protocol/0056-REWA-rewards_overview.md [hunk 18]
```diff
@@ -403,6 +403,14 @@ Create 3 markets settling in USDT. Wait for a new epoch to begin, in the next ep
...
+- Balance of the governance asset in the [Locked_FOR_REWARD] account counts towards the party's staking balance and can be nominated to a validator and earn staking rewards. (<a name="0056-REWA-202" href="#0056-REWA-202">0056-REWA-202</a>).
```
Comment (cdummett) [inline protocol/0056-REWA-rewards_overview.md]:
```suggestion
- Total balance of the governance asset in the relevant vesting account counts towards the party's staking balance and can be nominated to a validator and earn staking rewards. (<a name="0056-REWA-202" href="#0056-REWA-202">0056-REWA-202</a>).
```

File: protocol/0057-TRAN-transfers.md [hunk 19]
```diff
@@ -6,7 +6,10 @@ These transfers are not to be confused with the internal concept of transfers wh
 Allowing users to initiate transfers allows for the following capabilities:
 
 - A user can transfer funds from a public key A to a public key B.
-- A user can transfer funds from and to a locked account used for staking (yet to be specified) [LOCKED_FOR_STAKING](./0059-STKG-simple_staking_and_delegating.md).
+- A user can transfer governance tokens from general account to another key's general account.
+- A user can transfer governance tokens from and to a locked account used for staking [LOCKED_FOR_STAKING](./0059-STKG-simple_staking_and_delegating.md).
```
Comment (cdummett) [inline protocol/0057-TRAN-transfers.md:10]: Don't think we need this any more? Tokens should contribute automatically without transfer.

**Review by cdummett — CHANGES_REQUESTED** [review]

File: protocol/0057-TRAN-transfers.md [hunk 20]
```diff
@@ -262,8 +265,8 @@ message CancelTransfer {
...
-- As a user I can transfer funds from a general account I control to a locked_for_staking. Such transfer can be immediate or delayed. This functionality is currently not implemented (so don't try to test) (<a name="0057-PALAZZO-003" href="#0057-PALAZZO-003">0057-PALAZZO-003</a>).
-- As a user I can transfer funds from a locked_from_staking account under my control to any party's general_account. Such transfer can be immediate or delayed. This functionality is currently not implemented (so don't try to test) (<a name="0057-PALAZZO-004" href="#0057-PALAZZO-004">0057-PALAZZO-004</a>)
+- As a user I can transfer the governance token from a general account I control to a LOCKED_FOR_STAKING account. Such transfer must be immediate. (<a name="0057-TRAN-003" href="#0057-TRAN-003">0057-TRAN-003</a>).
+- As a user I can transfer the governance token from a LOCKED_FOR_STAKING account under my control to any party's general account. Such transfer must be immediate.  (<a name="0057-TRAN-004" href="#0057-TRAN-004">0057-TRAN-004</a>)
```
Comment (cdummett) [inline protocol/0057-TRAN-transfers.md:269]: Think these need to be added to features.json as well.

File: protocol/features.json [hunk 21]
```diff
@@ -679,7 +679,9 @@
       "0056-REWA-216",
       "0056-REWA-217",
       "0056-REWA-218",
-      "0056-REWA-219"
+      "0056-REWA-219", 
+      "0056-REWA-220", 
+      "0056-REWA-221"
```
Comment (cdummett) [inline protocol/features.json]: These need to go in a new feature section, otherwise they will come under reward scaling.

**Review by cdummett — CHANGES_REQUESTED** [review]

**Review by cdummett — APPROVED** [review]

**Pull request merged**